custom/icds_reports/utils/aggregation_helpers/distributed/agg_dashboard_acitivity.py

```diff
@@ -0,0 +1,272 @@
+                                                   }
+        return transformed_locations
+
+    def get_user_locations(self):
+        user_locations = list()
+        for user in self.get_dashboard_users:
+
+            state_id, district_id, block_id, user_level = None, None, None, None
+
+            if user['location_id']:
+                user_level = self.transformed_locations.get(user['location_id'])['loc_level']
+
+                if user_level == 1:
+                    state_id = user['location_id']
+                    district_id = 'All'
+                    block_id = 'All'
+                elif user_level == 2:
+                    state_id = self.transformed_locations.get(user['location_id'])['parents']['state_id']
+                    district_id = user['location_id']
+                    block_id = 'All'
+                elif user_level == 3:
+                    state_id = self.transformed_locations.get(user['location_id'])['parents']['state_id']
+                    district_id = self.transformed_locations.get(user['location_id'])['parents']['district_id']
+                    block_id = user['location_id']
+
+            user_locations.append((
+                user['username'],
+                state_id,
+                district_id,
+                block_id,
+                user_level
+            ))
+
+        return user_locations
+
+    @property
+    def tablename(self):
+        return "{}_{}".format(self.aggregate_parent_table, self.date.strftime("%Y-%m-%d"))
+
+    def drop_table_query(self):
+        return 'DROP TABLE IF EXISTS "{}"'.format(self.tablename)
+
+    def create_table_query(self):
+        return """
+        CREATE TABLE IF NOT EXISTS "{tablename}" (
+            CHECK (date = DATE %(date)s),
+            LIKE "{parent_tablename}" INCLUDING DEFAULTS INCLUDING CONSTRAINTS INCLUDING INDEXES
+        ) INHERITS ("{parent_tablename}")
+        """.format(
+            parent_tablename=self.aggregate_parent_table,
+            tablename=self.tablename,
+        ), {
+            "date": self.date.strftime("%Y-%m-%d"),
+        }
+
+    def get_last_agg_date(self):
+        from custom.icds_reports.models.aggregate import DashboardUserActivityReport
+        result = DashboardUserActivityReport.objects.\
+            filter(date__lt=self.date.strftime("%Y-%m-%d")).\
+            aggregate(Max('date'))
+        if result:
+            return result['date__max']
+        return None
+
+    def add_missing_users(self):
+
+        parameters = {'value{}'.format(index): tuple(list(loc) + [self.date])
+                      for index, loc in enumerate(self.get_user_locations())
+                      }
+        param_keys = ['%({})s'.format(param) for param in parameters.keys()]
+
+        return """
+        INSERT INTO "{tablename}" (
+            username, state_id, district_id,block_id,
+            user_level,date
+        )
+        VALUES {param_keys}
```

Comment:
postgres docs imply that the values here should always be wrapped in parens, why is that unnecessary here? and can you show what this query looks like? i am having a hard time visualizing through the multiple layers of parameter substitution. Is it possible to make the values more explicit?

Reply:
Query will be like this
```
INSERT INTO "dashboard_user_activity_2019-12-04" (
            username, state_id, district_id,block_id,
            user_level,date
        )
        VALUES %(value0)s,%(value1)s,%(value2)s

```

And the params will be

```
{'value0': ('23.test@icds-cas.commcarehq.org', 'st1', 'd1', 'All', 2, datetime.date(2019, 12, 4)), 'value1': ('34.test@icds-cas.commcarehq.org', 'st1', 'd1', 'b1', 3, datetime.date(2019, 12, 4)), 'value2': ('12.test@icds-cas.commcarehq.org', 'st1', 'All', 'All', 1, datetime.date(2019, 12, 4))}
```
@calellowitz

Reply:
and the final query which runs on db will be like this https://stackoverflow.com/a/20820950/4215264